clang/lib/Sema/BoundsAnalysis.cpp

```diff
@@ -83,13 +87,76 @@ void BoundsAnalysis::InitInOutSets() {
   }
 }
 
+llvm::APSInt BoundsAnalysis::GetSwitchCaseVal(const Expr *CaseExpr) {
+  Expr *E = const_cast<Expr *>(CaseExpr);
+
+  while (auto *ICE = dyn_cast<ImplicitCastExpr>(E))
+    E = ICE->getSubExpr();
+
+  llvm::APSInt IntVal (Ctx.getTypeSize(CaseExpr->getType()), 0);
+
+  // case '1': This is how it is represented in the CFG:
+  // ConstantExpr 'int'
+  //   -CharacterLiteral 'int' 49
+  // This is an IntegerConstantExpr. After the call to isIntegerConstantExpr,
+  // the variable IntVal would contain the value '1'.
+  if (E->isIntegerConstantExpr(IntVal, Ctx))
+    return IntVal;
+
+  // const int i = 1;
+  // case i: This is how it is represented in the CFG:
+  // ConstantExpr 'const int' lvalue 1
+  //   -DeclRefExpr 'const int' lvalue Var 'i' 'const int'
+  // This is not an IntegerConstantExpr. It is a ConstantExpr and we get its
+  // value by calling the getResultAsAPSInt() method.
+
+  // Note: According to C11 spec sections 6.6 and 6.8.1 a DeclRefExpr is not
+  // considered a constant expr. But clang allows for this additional
+  // extension. So we handle this here.
+  if (const auto *CE = dyn_cast_or_null<ConstantExpr>(E))
```

Comment:
Consider this case:
`const int i = 1;`

This is represented in AST as:
`ConstantExpr 'const int' lvalue 1`
`  -- DeclRefExpr 'const int' lvalue Var 'i' 'const int'`

The case for ConstantExpr invokes CheckICE on the DeclRefExpr. Now DeclRefExprs only seem to be considered ICE for EnumConstantDecl or when the language is C++.

```

That's the reason I had a separate case for ConstantExpr.
case Expr::DeclRefExprClass: {
     if (isa<EnumConstantDecl>(cast<DeclRefExpr>(E)->getDecl()))
       return NoDiag();
     const ValueDecl *D = cast<DeclRefExpr>(E)->getDecl();
     if (Ctx.getLangOpts().CPlusPlus &&
         D && IsConstNonVolatile(D->getType())) {
     ...
     }
     return ICEDiag(IK_NotICE, E->getBeginLoc());
```

Reply:
Thank you for clarifying this and giving this concrete example.  Yes, it makes sense to keep this case.   In test cases where this non-standard extension is used, could you call out that the usage is non-standard C, but allowed by clang.

Reply:
Thanks David. 
And as a result of keeping this case we also need to strip off casts because `getResultAsAPSInt` does not recursively test subexprs.

```
llvm::APSInt ConstantExpr::getResultAsAPSInt() const {
  switch (ConstantExprBits.ResultKind) {
  case ConstantExpr::RSK_APValue:
    return APValueResult().getInt();
  case ConstantExpr::RSK_Int64:
    return llvm::APSInt(llvm::APInt(ConstantExprBits.BitWidth, Int64Result()),
                        ConstantExprBits.IsUnsigned);
  default:
    llvm_unreachable("invalid Accessor");
  }
}
```

I can move the stripping off casts to just above the call to `getResultAsAPSInt`.

Reply:
It makes sense to move stripping off casts earlier.   I am still concerned that stripping off casts could cause our analysis to differ from what  being tested at runtime.   What does the code generation phase do for this case?  How are they handling casts?

Reply:
I meant "move stripping off casts later".

Reply:
The overflown value is truncated to the valid integer range in clang ("lib/Sema/SemaStmt.cpp"):
```
    llvm::APSInt ConvVal(Val);
    AdjustAPSInt(ConvVal, UnpromotedWidth, UnpromotedSign);
    AdjustAPSInt(ConvVal, Val.getBitWidth(), Val.isSigned());
    if (ConvVal != Val)
      S.Diag(Loc, diag::warn_case_value_overflow);

static void AdjustAPSInt(llvm::APSInt &Val, unsigned BitWidth, bool IsSigned) {
  Val = Val.extOrTrunc(BitWidth);
  Val.setIsSigned(IsSigned);
}
```

So in the AST there is no cast for the trunc/ext as the AST does not record the trunc/ext.
And the codegen simply receives the trunc/ext value. Here's the codegen for the following reduced test case:
```
void f(unsigned int a) {
  switch (a) {
  case 0x500000000LL:
    break;
  }
}
```
Codegen:
```
        sub     sp, sp, #4
        str     r0, [sp]
        ldr     r0, [sp]
        cmp     r0, #0
        bne     .LBB0_2
        b       .LBB0_1
.LBB0_1:                                @ %sw.bb
        b       .LBB0_2
.LBB0_2:                                @ %sw.epilog
        add     sp, sp, #4
```

As we can see the value of the label (cmp r0, #0) is 0.